tests/integration/security/util/reachability/context.go

```diff
@@ -64,16 +66,19 @@ type Context struct {
 	Multiversion echo.Instance
 	Headless     echo.Instance
 	Naked        echo.Instance
+	VM           echo.Instance
 }
 
 // CreateContext creates and initializes reachability context.
-func CreateContext(ctx framework.TestContext, p pilot.Instance) Context {
+func CreateContext(ctx framework.TestContext, p pilot.Instance, buildVM bool) Context {
```

Comment:
Why not always create the VM in the test?

Reply:
Thanks for your comment! Because tests in other packages might use this to create context, and the creation would fail if the Istio does not have `meshExpansion` enabled. If there's a better way other than passing an option to create the VM, I would be happy to update it!

Reply:
Why not enable meshExpansion everywhere? If the meshExpansion is not supposed to be enabled by default, it should have its own setup instead of plugging into the reachability_test.go and change the default install option for all other tests. The tests could miss a bug that is only triggered when meshExpansion is disabled.

Similar things have been done to other tests, for example, the mtlsk8sca/strict_test.go has its own install option `cfg.Values["global.pilotCertProvider"] = "kubernetes"`, the mtls_first_party_jwt/strict_test.go has `cfg.Values["global.jwtPolicy"] = "first-party-jwt"`, the filebased_tls_origination has its own special `istio-egressgateway` configuration.

Reply:
Thanks for your comment! I agree that we should not enable meshExpansion everywhere. We should enable it only when we need it. I think I am doing what you described [here](https://github.com/istio/istio/blob/a4f223b39537739fdd8d9ac191044824bf60bb57/tests/integration/security/main_test.go#L46-L62), so I am a little confused about what you are suggesting. The meshExpansion is currently enabled only for tests in the `security` package. mtlsk8sca/strict_test.go will not have the `meshExpansion` option enabled, and since it uses CreateContext() function to prepare a test context, I used a bool to make sure it will not try to create a VM. Could you please kindly elaborate further regarding what changes should I make? Thank you!

Reply:
It's also enabled for authorization_test.go, jwt_test.go and pass_through_filter_chain_test.go and mtls_healthcheck_test.go, my concern is it reduces the test coverage because the meshExpansion is not a default-on config for those tests.

Reply:
That's indeed a valid concern. The point I am doing this is 1) we don't want to duplicate the tests to create tests solely for VMs (they share the basic logic and the test structure) 2) we will be extending more tests to support VM testing in different areas (pilot, security, telemetry, etc.) and it might be the case that the other tests you just mention would be extended as well. 

We actually had the same conversation with @maasen , @howardjohn  and @elfinhe . Could you please chime in here so that we could reach a consensus about whether or not we should create dedicated Istio installation for these VM tests? If yes, I would create a separate package for these extended tests so that they could share the Istio installation?

Reply:
Please don't make a new install... meshExpansion being enabled doesn't make all those tests you mentioned less valid... just so we are on the same page - the flag simply opens a few more ports on the ingress gateway. I don't see how that can impact the other tests and give us less test coverage, at least not worth an extra 10 min of testing

Reply:
If the meshExpansion is just opening a few more ports on the ingress gateway, then why `the creation would fail if the Istio does not have meshExpansion enabled` as mentioned by @ZhengzheYang ? shouldn't it just work for all tests in this case.

All it does here in the CreateContext() is to deploy an extra VM instance which don't know why it needs to be controlled by a buildVM flag. And if that doesn't make much difference in the code path in pilot, we should just enable them for all other tests as well, instead of doing it only in reachability_test.go.

Reply:
The other tests fail simply because they couldn't find a port for the VM, since VM is an external workload. The test fail not because it's not passing the test, but because it couldn't create that VM, and the test would abort due to that. Hope this clarify the goal of `buildVM` flag here.

Reply:
I think I'm getting closer but still a little bit confused, so basically you have the `meshExpansion: true` in the config setup which according to Howard it is for opening some more ports on ingress gateway, and the change in this CreateContext() is setting `DeployAsVM: buildVM` to create the instance as VM (if the buildVM is true). How are these 2 things related?

Reply:
If the `DeployAsVM` is set to be true, it will require the `meshExpansion` to be turned on (Context here about deploying echo instance as VM: [link](https://github.com/istio/istio/pull/24166)). Therefore, in reachability.go test, `meshExpansion`  is on, so the creation of the VM would succeed. However, in tests like mtlsk8sca/strict_test.go, they use CreateContext() as well. If it also tries to create the vm instance with `DeployAsVM` as true, then it would fail, since the `meshExpansion` is off (there's no port for this echo instance as it's simulating an external workload). Hope this is a little clearer! :)

Reply:
1. I looked at the PR and it's not obvious but I guess I'm figuring this out: the flag `meshExpansion` exposes the istiod discovery port on ingressgsteway which is used by the VM instance to get the xDS response from istiod. Is this what you mean there is no port for this echo instance? Maybe @howardjohn has more context.

2. Then it goes to my other question in the beginning that why not just set `meshExpansion: true` for all other tests? So that we don't need to specify this `buildVM` and can just always test the VM instance.

Reply:
Thanks for your comment! To answer your second question, I guess we could, but that would involve a lot of code changes and a huge PR, and a lot of people should be part of the code reviewing process. I am not the author of the most of the tests so that would put a burden on both me and the code owners. For now, our way of covering the VM tests is extend the tests gradually, with minimal code changes. PRs like #24757 and #24879 are such examples. I am making PRs under a mindset that I should propose as few code changes as possible. 

I think I could enable the `meshExpansion` for the tests that call `CreateContext()` function, since that would eliminate the need to have this buildVM flag and build VM all the time, but that would introduce more complexities for the test cases, since the calls from `naked` instances to another external workload that is not injected with sidecar would fail, and we need to exclude those cases, which will introduce additional code changes. If you think this is preferable, I am happy to make the update and request a review from you again!

Reply:
no, the other tests I mean those you're modifying in this PR: `cni/cni_test.go`, `mtls_first_party_jwt/strict_test.go` and `mtlsk8sca/strict_test.go`, so you basically can just update their TestMain to set the meshExpansion to true.


>  the calls from naked instances to another external workload that is not injected with sidecar would fail

How? The VM instance has its sidecar (https://github.com/istio/istio/blob/365a94720d5a5290fafeaee85d49ad83f144be1e/pkg/test/framework/components/echo/kube/deployment.go#L218) so shouldn't it behave similarly to other normal pods? And I think you only need 1 or 2 line code of change to exclude some invalid test cases.

Reply:
Thank you for reviewing my code! I will update and let you know when I am done patching!

Reply:
Hey Yangmin, I took your advice and set `meshExpansion` to be true for the other tests, but the tests failed because VM might not work with `cfg.Values["global.jwtPolicy"] = "first-party-jwt"` and `cfg.Values["global.pilotCertProvider"] = "kubernetes"`. In practice, according to the [doc](https://istio.io/latest/docs/examples/virtual-machines/single-network/#preparing-the-kubernetes-cluster-for-vms) about VM onboarding, users are the ones that are responsible for preparing for the certs for the VMs. For the current setup of the tests, VM might not work with first-party-jwt and certs provided by kube. Therefore, for now I think it might be reasonable to keep the `buildVM` parameter so that VM is not also deployed as VMs for the other tests. Does this sound reasonable to you?

Reply:
I see, there might be some gaps there, could you open an issue for tracking purpose? thanks.
cc @lei-tang